api/documents/query.go

```diff
@@ -0,0 +1,169 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package documents
+
+import (
+	"context"
+	"fmt"
+
+	v1 "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+)
+
+// Query - Query interface for Document Service
+type Query interface {
+	// Where - Append an expression to the query
+	Where(...*queryExpression) Query
+
+	// Limit - limit the max result size of the query
+	Limit(int) Query
+	// FromPagingToken - Start from a given entry
+	FromPagingToken(interface{}) Query
+
+	// Fetch - Return paged values
+	Fetch() (*FetchResult, error)
+
+	// Stream - Return an iterator containing values
+	Stream() (DocumentIter, error)
+}
+
+// Defacto Query interface implementation
+type queryImpl struct {
+	col   CollectionRef
+	dc    v1.DocumentServiceClient
+	pt    interface{}
+	exps  []*queryExpression
+	limit int
+}
+
+func (q *queryImpl) Where(qes ...*queryExpression) Query {
+	q.exps = append(q.exps, qes...)
+
+	return q
+}
+
+func (q *queryImpl) Limit(limit int) Query {
+	q.limit = limit
+	return q
+}
+
+func (q *queryImpl) FromPagingToken(token interface{}) Query {
+	q.pt = token
+	return q
+}
+
+type FetchResult struct {
+	Documents   []Document
+	PagingToken interface{}
+}
+
+func (q *queryImpl) Fetch() (*FetchResult, error) {
+	// build the expressions list
+	expressions := make([]*v1.Expression, 0, len(q.exps))
+
+	for _, e := range q.exps {
+		wexp, err := e.toWire()
+
+		if err != nil {
+			return nil, err
+		}
+
+		expressions = append(expressions, wexp)
+	}
+
+	var token map[string]string = nil
+	if q.pt != nil {
+		t, ok := q.pt.(map[string]string)
+
+		if !ok {
+			return nil, fmt.Errorf("Invalid paging token provided!")
+		}
+		token = t
+	}
+
+	r, err := q.dc.Query(context.TODO(), &v1.DocumentQueryRequest{
+		Collection:  q.col.toWire(),
+		Expressions: expressions,
+		Limit:       int32(q.limit),
+		PagingToken: token,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	docs := make([]Document, 0, len(r.GetDocuments()))
+
+	for _, d := range r.GetDocuments() {
+
+		fmt.Println("testing", d.GetKey())
+		// TODO: Handle error
+		// For now we'll just set the back reference to nil
+		ref, err := documentRefFromWireKey(q.dc, d.GetKey())
+
+		if err != nil {
+			// XXX: Potentially just log an error and continue
+			return nil, fmt.Errorf("Fetch Error: Failed to construct document reference for result (%v)", err)
+		}
+
+		docs = append(docs, &documentImpl{
+			ref:     ref,
+			content: d.Content.AsMap(),
+		})
+	}
+
+	return &FetchResult{
+		Documents:   docs,
+		PagingToken: r.GetPagingToken(),
+	}, nil
+}
+
+func (q *queryImpl) Stream() (DocumentIter, error) {
+	// build the expressions list
+	expressions := make([]*v1.Expression, 0, len(q.exps))
```

Comment:
This is repeated code from Fetch, should be extracted

Reply:
done